File: cpp/structure/decorator.cpp
#include <iostream>
using namespace std;
class IA {
public:
    virtual void f() = 0;
};
class A1: public IA {
public:
    void f() {
        cout<<"A1 process" <<endl;
    }
};

class A2: public IA {
public:
    void f() {
        cout<<"A2 process" <<endl;
    }
};
class ADecorator: public IA {
private:
    IA* ia;
public:
    ADecorator(IA* ia_):ia(ia_) { }
    void f() {
        cout<<"ADecorator process" <<endl;
        ia->f();
    }
};
int main() {
    IA* i1 = new A1();
    IA* i2 = new A2();
    IA* d1 = new ADecorator(i1);
    d1->f();

    IA* d2 = new ADecorator(d1);
    d2->f();
}

File: golang/structure/decorator.go
package main

type IA interface {
	f()
}

type A1 struct {
}
func (a *A1) f() {
	println("A1 process")
}

type A2 struct {
}
func (a *A2) f() {
	println("A2 process")
}

type Decorator struct {
	ia IA
}
func NewDecorator(ia IA) IA {
	return &Decorator{ia}
}
func (d *Decorator) f() {
	println("Decorator process")
	d.ia.f()
}

func main() {
	a1 := A1{}
	a2 := A2{}
	d1 := NewDecorator(&a1)
	d1.f()

	d2 := NewDecorator(&a2)
	d2.f()

	d3 := NewDecorator(d1)
	d3.f()
}
